src/src/com/microsoft/aad/adal/TokenCacheAccessor.java

```diff
@@ -0,0 +1,271 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.aad.adal;
+
+import java.io.UnsupportedEncodingException;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Internal class handling the interaction with {@link AcquireTokenSilentHandler} and {@link ITokenCacheStore}. 
+ */
+class TokenCacheAccessor {
+    private static final String TAG = TokenCacheAccessor.class.getSimpleName();
+    
+    private final ITokenCacheStore mTokenCacheStore;
+    private final String mAuthority;
+    
+    TokenCacheAccessor(final ITokenCacheStore tokenCacheStore, final String authority) {
+        if (tokenCacheStore == null) {
+            throw new IllegalArgumentException("tokenCacheStore");
+        }
+        
+        if (StringExtensions.IsNullOrBlank(authority)) {
+            throw new IllegalArgumentException("authority");
+        }
+        
+        mTokenCacheStore = tokenCacheStore;
+        mAuthority = authority;
+    }
+    
+    /**
+     * @return {@link TokenCacheItem} for regular token cache entry.  
+     */
+    TokenCacheItem getRegularTokenCacheItemWithAT(final String resource, final String clientId, final String user) {
+        final String cacheKey = CacheKey.createCacheKeyForRTEntry(mAuthority, resource, clientId, user);
+        return mTokenCacheStore.getItem(cacheKey);
+    }
+    
+    /**
+     * @return {@link TokenCacheItem} for MRRT token cache entry.  
+     */
+    TokenCacheItem getMRRTItem(final String clientId, final String user) {
+        final String cacheKey = CacheKey.createCacheKeyForMRRT(mAuthority, clientId, user);
+        return mTokenCacheStore.getItem(cacheKey);
+    }
+    
+    /**
+     * @return {@link TokenCacheItem} for FRT token cache entry.  
+     */
+    TokenCacheItem getFRTItem(final String familyClientId, final String user) {
+        if (StringExtensions.IsNullOrBlank(user)) {
+            return null;
+        }
+        
+        final String cacheKey = CacheKey.createCacheKeyForFRT(mAuthority, familyClientId, user);
+        return mTokenCacheStore.getItem(cacheKey);
+    }
+    
+    /**
+     * Remove existing token cache item if needed and update token cache with returned auth result. 
```

Comment:
Suggestion: I think this function should be refactored into 2 parts.

---
In reply to: [62960934](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/611#discussion_r62960934) [](ancestors = 62960934)